wandb/sdk/internal/job_builder.py

```diff
@@ -0,0 +1,252 @@
+"""
+job builder.
+"""
+import json
+import os
+import sys
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union
+
+from wandb.sdk.data_types._dtypes import TypeRegistry
+from wandb.sdk.lib.filenames import DIFF_FNAME, METADATA_FNAME, REQUIREMENTS_FNAME
+from wandb.sdk.wandb_artifacts import Artifact
+from wandb.util import make_artifact_name_safe
+
+from .settings_static import SettingsStatic
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    from typing_extensions import TypedDict
+
+if TYPE_CHECKING:
+    from wandb.proto.wandb_internal_pb2 import ArtifactRecord
+
+FROZEN_REQUIREMENTS_FNAME = "requirements.frozen.txt"
+JOB_FNAME = "wandb-job.json"
+JOB_ARTIFACT_TYPE = "job"
+
+
+class GitInfo(TypedDict):
+    remote: str
+    commit: str
+
+
+class GitSourceDict(TypedDict):
+    git: GitInfo
+    entrypoint: List[str]
+    args: List[str]
+
+
+class ArtifactSourceDict(TypedDict):
+    artifact: str
+    entrypoint: List[str]
+    args: List[str]
+
+
+class ImageSourceDict(TypedDict):
+    image: str
+    args: List[str]
+
+
+class JobSourceDict(TypedDict, total=False):
+    _version: str
+    source_type: str
+    source: Union[GitSourceDict, ArtifactSourceDict, ImageSourceDict]
+    input_types: Dict[str, Any]
+    output_types: Dict[str, Any]
+    runtime: Optional[str]
+
+
+class ArtifactInfoForJob(TypedDict):
+    id: str
+    name: str
+
+
+class JobBuilder:
+    _settings: SettingsStatic
+    _metadatafile_path: Optional[str]
+    _requirements_path: Optional[str]
+    _config: Optional[Dict[str, Any]]
+    _summary: Optional[Dict[str, Any]]
+    _logged_code_artifact: Optional[ArtifactInfoForJob]
+    _used_job: bool
+
+    def __init__(self, settings: SettingsStatic):
+        self._settings = settings
+        self._metadatafile_path = None
+        self._requirements_path = None
+        self._config = None
+        self._summary = None
+        self._logged_code_artifact = None
+        self._used_job = False
+
+    def set_config(self, config: Dict[str, Any]) -> None:
+        self._config = config
+
+    def set_summary(self, summary: Dict[str, Any]) -> None:
+        self._summary = summary
+
+    @property
+    def used_job(self) -> bool:
+        return self._used_job
+
+    def set_used_job(self, val: bool) -> None:
+        self._used_job = val
+
+    def _set_logged_code_artifact(
+        self, res: Optional[Dict], artifact: "ArtifactRecord"
+    ) -> None:
+        if artifact.type == "code" and res is not None:
+            self._logged_code_artifact = ArtifactInfoForJob(
+                {
+                    "id": res["id"],
+                    "name": artifact.name,
+                }
+            )
+
+    def _build_repo_job(
+        self, remote: str, commit: str, program_relpath: str, args: List[str]
+    ) -> Tuple[Artifact, GitSourceDict]:
+        source: GitSourceDict = {
+            "entrypoint": [
+                os.path.basename(sys.executable),
```

Comment:
is this always the executable? we had some issues with people that use pex and other different setups

Reply:
Interesting, as far as I can tell, `sys.executable` is still being used when we track the executable in `system_info.py` is there another pattern I should be using?

Reply:
good question, i think we were trying to add this field to the settings object under `_executable`

although unrelated the pex fix seems to be broken, i have a PR that revert it to something more standard and allows users to specify any "exotic" executable (#4604)

Reply:
I'll leave this as is for now until we have a better way to check the executable. This works for all current users. I'll put a TODO